### PR TITLE
WebDriver+WebContent+LibWeb: Progress towards clicking

### PR DESCRIPTION
**Pull request by AtkinsSJ**
A very small step towards making "element click" work, but I realised that further progress requires a bunch of infrastructure we're missing. (And possibly rearranging how our webdriver code is split across processes. :grimacing:)

### CODE REVIEW

File: Userland/Services/WebContent/WebDriverConnection.cpp [hunk 1]
```diff
@@ -1272,8 +1289,90 @@ Messages::WebDriverClient::ElementClickResponse WebDriverConnection::element_cli
     // FIXME: 7. If element’s container is obscured by another element, return error with error code element click intercepted.
 
     // 8. Matching on element:
-    // FIXME: option element
-    // FIXME: Otherwise
+    //    - option element
```
Comment (linusg) [inline Userland/Services/WebContent/WebDriverConnection.cpp:1292]: No indentation here, IMO

File: Userland/Services/WebContent/WebDriverConnection.cpp [hunk 2]
```diff
@@ -1272,8 +1289,90 @@ Messages::WebDriverClient::ElementClickResponse WebDriverConnection::element_cli
...
+                fire_an_event<Web::DOM::Event>("change", parent_node);
+            }
+        }
+        // 7. Fire a mouseUp event at parent node.
+        fire_an_event<Web::UIEvents::MouseEvent>("mouseUp", parent_node);
+
+        // 8. Fire a click event at parent node.
+        fire_an_event<Web::UIEvents::MouseEvent>("click", parent_node);
+    }
+    //    - Otherwise
```
Comment (linusg) [inline Userland/Services/WebContent/WebDriverConnection.cpp:1340]: Same

File: Userland/Services/WebContent/WebDriverConnection.cpp [hunk 3]
```diff
@@ -1272,8 +1289,90 @@ Messages::WebDriverClient::ElementClickResponse WebDriverConnection::element_cli
...
+    else {
+        // FIXME: 1. Let input state be the result of get the input state given current session and current top-level browsing context.
```
Comment (linusg) [inline Userland/Services/WebContent/WebDriverConnection.cpp:1342]: I'd just say "FIXME: Implement steps 1-17", but we can also do this :sweat_smile:

File: Userland/Libraries/LibWeb/DOM/NodeFilter.h [hunk 4]
```diff
@@ -27,6 +27,19 @@ class NodeFilter final : public Bindings::PlatformObject {
         FILTER_SKIP = 3,
     };
 
+    enum WhatToShow {
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibWeb/DOM/NodeFilter.h:30]: If this is just a plain `enum` for bitwise operations, we have `AK_ENUM_BITWISE_OPERATORS` for that.

[the remark this answers is not in the document]
Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/DOM/NodeFilter.h:30]: As a side note, is there some way of not having to define these here at all? They're in the IDL file but I don't think that ends up exposed to C++.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibWeb/DOM/NodeFilter.h:30]: Nope, these are translated like this:

```cpp
auto constant_SHOW_ALL_value = JS::Value((u32)0xFFFFFFFF);
define_direct_property("SHOW_ALL", constant_SHOW_ALL_value, JS::Attribute::Enumerable);
auto constant_SHOW_ELEMENT_value = JS::Value((u32)0x1);
define_direct_property("SHOW_ELEMENT", constant_SHOW_ELEMENT_value, JS::Attribute::Enumerable);
auto constant_SHOW_ATTRIBUTE_value = JS::Value((u32)0x2);
define_direct_property("SHOW_ATTRIBUTE", constant_SHOW_ATTRIBUTE_value, JS::Attribute::Enumerable);
...
```

File: Userland/Services/WebContent/WebDriverConnection.cpp (hunk 3 above)
Comment (AtkinsSJ) [reply Userland/Services/WebContent/WebDriverConnection.cpp:1342]: Originally I was actually going to implement them but I ran out of steam. :sweat_smile:

File: Userland/Libraries/LibWeb/DOM/NodeFilter.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/DOM/NodeFilter.h:30]: Hmmm. Something to add to the yak-stack then.

The NAMES_LIKE_THIS isn't required but seems to be how all lists-of-constants-from-IDL are named, to match the IDL definitions. But yeah long-term I'd like these enums to not exist and just be generated from IDL. (Though, making them be enums might be tricky.)

**Comment by AtkinsSJ** [comment]
Handled all feedback.

**Pull request merged**